### PR TITLE
refactor SignatureHash

### PR DESCRIPTION
**Pull request by dudebing99**

### CODE REVIEW

File: src/tx.h [hunk 1]
```diff
@@ -179,7 +180,7 @@ class CBaseTx {
     virtual uint256 GetHash() const = 0;
     virtual uint64_t GetFee() const = 0;
     virtual double GetPriority() const = 0;
-    virtual uint256 SignatureHash() const = 0;
+    virtual uint256 SignatureHash(bool reCalculate = false) const = 0;
```
Comment (chenxinhua2018) [inline src/tx.h]: "reCalculate" should be a word，write as "recalculate"?

File: src/tx.h [hunk 2]
```diff
@@ -243,12 +244,20 @@ class CRegisterAccountTx : public CBaseTx {
...
     uint64_t GetFee() const { return llFees; }
-    uint256 SignatureHash() const {
-        CHashWriter ss(SER_GETHASH, 0);
-        CID userPubkey(userId);
-        CID minerPubkey(minerId);
-        ss << VARINT(nVersion) << nTxType << VARINT(nValidHeight) << userPubkey << minerPubkey << VARINT(llFees);
-        return ss.GetHash();
+    uint256 SignatureHash(bool reCalculate = false) const {
+        static std::atomic_bool init{false};
+        static uint256 hash;
```
Comment (chenxinhua2018) [inline src/tx.h]: can not be static global variable here,

File: src/tx.h (hunk 1 above)
Comment (dudebing99) [reply src/tx.h]: done

File: src/tx.h (hunk 2 above)
Comment (dudebing99) [reply src/tx.h]: done

**Pull request merged**